### DISCUSSION TITLE
Add `DTO`

### DISCUSSION THREAD

**Discussion by Mohammad-Alavi** [apiato/apiato#709] [Ideas]
Since removing [Transporters](https://apiato.io/docs/9.x/main-components/transporters) for various reasons, I tried to fill its place with another DTO implementation.

For now the best candidate is [Laravel-data](https://spatie.be/docs/laravel-data/v4/introduction) package by spatie IMO.
But, to use this package we have to change some of the ways we do things right now, namely:
1. validation responsibility will be moved from Request to DTO
https://spatie.be/docs/laravel-data/v3/advanced-usage/validation-attributes
2. We can continue using the current Transformers or move to use laravel-data transforming style of the resource (I prefer our current transformers)
https://spatie.be/docs/laravel-data/v3/as-a-resource/from-data-to-resource

Your thoughts are important on this. 
There are a lot of pros in using DTO:
1. Actions wont be dependent on Request
2. Because the validation is in the DTO we can always be sure we are getting the right data when we need it anywhere in the system e.g. in Tasks
3. We will leverage a lot of typing features of the PHP 8.0 (while creating a DTO) which prevents a lot of errors and checks down the line
4. We wont see methods with looooong parameter list anymore
5. It would be much easier to test Actions  

Tell me about the cons. I can't think of many atm.
The only con I can think of for now is that it adds another package/documentation for us and newcomers to read/learn 👀

**Comment by mderis** [comment]
I appreciate the discussion around using DTOs to pass data from controllers to action classes in Apiato. As many of us know, using Request classes to validate data, authorize users, and pass data to actions seems convenient but can lead to issues in larger projects with complex data structures and multiple developers. I agree that separating concerns by using DTOs for data transfer has clear benefits:

1. **Separation of Concerns**: Request classes should focus on validation and authorization, while DTOs manage data transfer, making the codebase cleaner and more maintainable.
2. **Consistency and Clarity**: DTOs provide a clear contract for data structures, helping developers understand the data flow and making it easier to manage changes.
3. **Scalability**: In larger projects, having a dedicated layer for data transfer can reduce coupling and improve the scalability of the application.
4. **Testability**: DTOs can be independently tested, enhancing the overall testability of the code.

While the benefits are substantial, I believe DTOs should be an optional enhancement rather than a mandatory change. This approach would allow developers to adopt DTOs at their own pace without causing any breaking changes. Here are a few reasons for keeping DTOs optional:

1. **Backward Compatibility**: Ensuring that existing projects continue to function without modification is crucial. By making DTOs optional, we respect current workflows while offering a path to improvement.
2. **Developer Choice**: Each project has unique needs and constraints. Providing flexibility allows teams to choose the best approach for their specific context.
3. **Gradual Adoption**: Teams can gradually integrate DTOs, starting with new features or modules, which can be a more manageable way to implement changes in a large codebase.

To implement this, I propose that we introduce DTOs in a way that does not disrupt existing functionalities. Documentation and examples can highlight the benefits and usage of DTOs, encouraging developers to consider this approach while preserving the traditional method as a valid option.

By striking a balance between innovation and flexibility, we can enhance Apiato’s robustness and maintain its usability for a wide range of projects and developer preferences.

**Reply by mderis** [reply]
I forgot to mention an important point:

One more benefit of using DTOs is that by not being dependent on Request classes, we can reuse each action file in API controllers, Web controllers, and jobs. 

Additionally, it is a better practice to test actions without being dependent on requests, leading to cleaner unit tests.

**Comment by i-prikot** [comment]
Look at this package https://github.com/WendellAdriel/laravel-validated-dto
